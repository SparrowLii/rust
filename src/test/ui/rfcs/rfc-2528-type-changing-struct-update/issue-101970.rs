#![feature(type_changing_struct_update)]
#![allow(incomplete_features)]

#[derive(Default)]
pub struct Foo<P, T> {
    pub t: T,
    pub v: P
}

impl<P: Default, T: Default> Foo<P, T> {
    pub fn new() -> Self {
        Foo { t: T::default(), v: P::default() }
    }

    pub fn d(t: T) -> Self {
        Foo { t, ..Default::default() }
    }

    pub fn o(t: T) -> Self {
        let d: Foo<_, _> = Foo::new();
        Foo { t, ..d }
    }

    pub fn o2<P2: Default>(t: T, v: P2) -> (Foo<P, T>, Foo<P2, T>) {
        let d = Default::default();
        let foo1 = Foo { t, ..d };
        let foo2 = Foo { v, ..d };
        (foo1, foo2)
    }

    pub fn o3<T2: Default>(t1: T, t2: T2) -> (Foo<P, T>, Foo<P, T2>) {
        let d = Default::default();
        let foo1 = Foo { t: t1, ..d };
        let foo2 = Foo { t: t2, ..d };
                                  //~^ ERROR mismatched types [E0308]
        (foo1, foo2)
    }
}

fn main() {}
