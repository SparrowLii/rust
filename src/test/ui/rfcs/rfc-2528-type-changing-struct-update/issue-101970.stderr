error[E0308]: mismatched types
  --> $DIR/issue-101970.rs:34:35
   |
LL | impl<P: Default, T: Default> Foo<P, T> {
   |                  - found type parameter
...
LL |     pub fn o3<T2: Default>(t1: T, t2: T2) -> (Foo<P, T>, Foo<P, T2>) {
   |               -- expected type parameter
...
LL |         let foo2 = Foo { t: t2, ..d };
   |                                   ^ expected type parameter `T2`, found type parameter `T`
   |
   = note: expected type parameter `T2`
              found type parameter `T`
   = note: a type parameter was expected, but a different one was found; you might be missing a type parameter or trait bound
   = note: for more information, visit https://doc.rust-lang.org/book/ch10-02-traits.html#traits-as-parameters

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
